### ISSUE TITLE
Vim-qf breaks vimtex

### ISSUE THREAD

**Issue by krishnakumarg1984** [romainl/vim-qf#69]
This plugin overrides vimtex's quickfix handler for parsing latex compilation errors and warnings.

**Comment by romainl** [comment]
Language-support plugins shouldn't do anything outside of their scope and generic plugins shouldn't have to do anything language-specific.

Two possibilities:

* disable vim-qf for TeX
* disable VimTeX's quickfix handling.

**Comment by lervag** [comment]
@romainl I agree, but this is a special case. In order to handle the hard-to-parse LaTeX log format I first parse it with an errorformat string. But to fix some paths, I use the `QuickFixCmdPost` autocommand. I think this is what is causing the problem here.

**Comment by romainl** [comment]
@lervag It's very likely that https://github.com/romainl/vim-qf/pull/73 fixed the issue. Do you use `:help autocmd-nested` in vimtex?

**Comment by lervag** [comment]
Interesting. No, I have not used nested autocmds. I'm aware of the feature, though, but I never realized that some of my autocommands could/should be nested.

Here is the relevant code that I think was conflicting:

```vim
"
" We use a temporary autocmd to fix some paths in the quickfix entry
"
if self.config.fix_paths
  let self.main = a:tex
  let self.root = b:vimtex.root
  augroup vimtex_qf_tmp
    autocmd!
    autocmd QuickFixCmdPost [cl]*file call b:vimtex.qf.fix_paths()
  augroup END
endif

execute (a:jump ? 'cfile' : 'cgetfile') fnameescape(a:log)

if self.config.fix_paths
  autocmd! vimtex_qf_tmp
endif
```

I don't think the `QuickFixCmdPost` should need to be nested. Am I wrong?

**Comment by lervag** [comment]
Addendum: the `fix_paths()` function does not do anything that I think should invoke any other autocmds. It only iterates the `getqflist()` to update some file paths.

**Comment by krishnakumarg1984** [comment]
@romainl  When ```vim-qf``` is enabled, it somehow suppresses the functionality of ```vimtex```. ```vimtex``` extensively uses the quickfix list for standard warnings & errors. I think (if that option is enabled) ```vimtex``` uses the output from another external program ```pplatex``` to populate the qf-list.  It relies on a client-server model to automatically open the qf-list upon completion.  None of this functionality works when ```vim-qf``` plugin is enabled.

**Comment by lervag** [comment]
@krishnakumarg1984 It would help to see a minimal example for this issue. Can you make one?

**Comment by lervag** [comment]
@romainl No, I use my own `ftplugin/qf.vim`. Based on the last comment by @krishnakumarg1984 I think my earlier comments are not so relevant. I think this must be something else. But it is hard to pinpoint without a good, minimal example.

**Comment by Konfekt** [comment]
If I recall correctly, I had to remove `:caddfile` or `:caddbuffer`
from
```vim
autocmd QuickFixCmdPost make,grep,grepadd,vimgrep,vimgrepadd,cscope,helpgrep nested redraw! | cwindow | wincmd p
```
to make the Quickfix Window reconcile with Vimtex.
Perhaps removing `:caddfile` or `:caddbuffer` from

https://github.com/romainl/vim-qf/blob/4026bbc36a37a89c9515da5875ebb5ff5b5c3091/plugin/qf.vim#L61

mitigates this issue?

**Comment by romainl** [comment]
Closing this as there is no resolution in sight. I stand by my position: filetype-specific plugins have no business handling the quickfix window.

**Comment by lervag** [comment]
I'm curious, @romainl: In the current version of VimTeX, I no longer rely on autocommands. I use the `caddfile` command with a properly defined `errorformat`, then I use `getqflist` and `setqflist` to make some adjustments to make the file paths of the quickfix entries better. (LaTeX log files are hard to parse with `errorformat`, and I currently don't know a perfect solution - what I do is sort of hacky, but it also seems to work quite well.)

The relevant code is [here](https://github.com/lervag/vimtex/blob/8178ade37a8143209f85f13170b908aad699083f/autoload/vimtex/qf/latexlog.vim#L101-L145). You wrote initially:

> Language-support plugins shouldn't do anything outside of their scope and generic plugins shouldn't have to do anything language-specific.

IMHO, I'm not going outside the language-support scope here. But please, if you have thoughts on the contrary, I would like to hear them.

**Comment by lervag** [comment]
By the way, from some simple tests it seems the `caddfile` after loading `vim-qf` forces a change of buffer. The default behaviour of `caddfile` is simply to add errors from a file to the list, it does not open the quickfix list or anything like that.

**Comment by romainl** [comment]
@lervag I must admit I have never written a single line of LaTex in my life and I have no practical knowledge of your plugin so I have no idea if your plugin and mine still clash.

The situation I'm in is that vim-qf relies on the `QuickFixCmdPost` event a lot. Opening the qf/loc window after any command that changed the qf/loc list is, after all, its core feature. When another plugin executes a command that triggers that event and then either does something based on that event or explicitly does something like `:cwindow` or `:lclose` all kinds of race conditions appear because there are two plugins trying to do the same thing. This is why I would prefer plugins that are not qf/loc centric to limit themselves to populating the qf/loc list, leaving the rest up to the user or, eventually, other plugins. This is also why I am not particularly interested in making adjustments to vim-qf because other plugins step out of line.

> The default behaviour of `caddfile` is simply to add errors from a file to the list, it does not open the quickfix list or anything like that.

The only commands that natively open the qf/loc are `:copen`, `:lopen`, `:cwindow`, and `:lwindow`. None of the other qf-related commands do and Vim-qf's raison d'être is precisely to open the qf/loc window after *any* of those non-opening qf commands is performed.

Yes, `:caddfile` doesn't open the qf window and the whole point of vim-qf is to change that.

> I use the `caddfile` command with a properly defined `errorformat`, then I use `getqflist` and `setqflist` to make some adjustments to make the file paths of the quickfix entries better.

I would consider `:help :caddexpr` as an alternative, as it lets you do all the massaging you need beforehand:

    function! ParseErrors(filename)
        let error_file = readfile(a:filename)
        <your parsing logic here>
        return parsed_output
        " output may look like classic unix-like output:
        " filename:23:45 Hello world!
        " or it may just be whatever you got from your file
        " and that you know can be parsed by efm but with
        " a few substitutions here and there like filenames
    endfunction
    execute "caddexpr " .. ParseErrors('./foobar')

That way, there is no need to manipulate the qf/loc list directly and less possible friction with other scripts.

**Comment by lervag** [comment]
> When another plugin executes a command that triggers that event and then either does something based on that event or explicitly does something like `:cwindow` or `:lclose` all kinds of race conditions appear because there are two plugins trying to do the same thing.

VimTeX does not use the `QuickFixCmdPost` event. It uses the following commands: `:cclose` and (optionally) `:cwindow`. Further, VimTeX uses `caddfile` and `errorformat` to parse log files for quickfix entries. Also, `qetqflist()` and `setqflist()` are used sometimes to manipulate the entries of the quickfix list (e.g. filter out uninteresting warnings or fix path related issues as mentioned).

I would expect these commands to be safe to use. And it seems in most cases they are; it seems the only thing that prevents VimTeX to work with vim-qf is this line:

https://github.com/romainl/vim-qf/blob/127257057c4befb8c1026a7ba7e07713af1fb5c8/plugin/qf.vim#L60

If I remove `caddfile`, it seems like VimTeX works with vim-qf.

> This is why I would prefer plugins that are not qf/loc centric to limit themselves to populating the qf/loc list, leaving the rest up to the user or, eventually, other plugins. This is also why I am not particularly interested in making adjustments to vim-qf because other plugins step out of line.

I would ask if you are sure that shooting the events also for `caddfile` is really appropriate. If you insist it is, then I would ask if there is any way to "opt out" this particular command? Or, perhaps better for you, if there is any way I could change things on my end that would work better.

> The only commands that natively open the qf/loc are `:copen`, `:lopen`, `:cwindow`, and `:lwindow`. None of the other qf-related commands do and Vim-qf's raison d'être is precisely to open the qf/loc window after _any_ of those non-opening qf commands is performed.

IMHO, it seems vim-qf does _much_ more than this. When I'm reading the README
file, it seems you are referring to this point:

> * (optional) open the location/quickfix window automatically after :make, :grep, :lvimgrep and friends if there are valid locations/errors

Thus, it seems perhaps the current way to make vim-qf work with VimTeX is to
add `let g:qf_auto_open_quickfix = 0`.

> I would consider `:help :caddexpr` as an alternative, as it lets you do all the massaging you need beforehand:

Thanks for the suggestion, but this does not allow me to use the `errorformat` for parsing the error file? Or am I missing something?

**Comment by romainl** [comment]
> > When another plugin executes a command that triggers that event and then either does something based on that event or explicitly does something like `:cwindow` or `:lclose` all kinds of race conditions appear because there are two plugins trying to do the same thing.
>
> VimTeX does not use the `QuickFixCmdPost` event. It uses the following commands: `:cclose` and (optionally) `:cwindow`. Further, VimTeX uses `caddfile` and `errorformat` to parse log files for quickfix entries. Also, `qetqflist()` and `setqflist()` are used sometimes to manipulate the entries of the quickfix list (e.g. filter out uninteresting warnings or fix path related issues as mentioned).

My opinion is that your plugin shouldn't use either `:cwindow` or `:cclose` or make any attempt whatsoever to manipulate the qf/loc window.

> 
> I would expect these commands to be safe to use. And it seems in most cases they are; it seems the only thing that prevents VimTeX to work with vim-qf is this line:
> 
> https://github.com/romainl/vim-qf/blob/127257057c4befb8c1026a7ba7e07713af1fb5c8/plugin/qf.vim#L60
> 
> If I remove `caddfile`, it seems like VimTeX works with vim-qf.
> 
> > This is why I would prefer plugins that are not qf/loc centric to limit themselves to populating the qf/loc list, leaving the rest up to the user or, eventually, other plugins. This is also why I am not particularly interested in making adjustments to vim-qf because other plugins step out of line.
> 
> I would ask if you are sure that shooting the events also for `caddfile` is really appropriate. If you insist it is, then I would ask if there is any way to "opt out" this particular command? Or, perhaps better for you, if there is any way I could change things on my end that would work better.

Here are a couple of reasons why keeping `:caddfile` in the list makes sense:

1. `:caddfile` behaves like `:cfile` when there is no existing list,
2. `:caddfile` potentially changes the length of the list, which in turn triggers vim-qf's auto-resizing feature,
3. again, the core feature of vim-qf is precisely to do things when the user executes any of the "quickfix command" listed under `:help QuickFixCmdPre`. User does `:grep`, or `:cexpr`, or `:caddfile`, or whatever => vim-qf does something.

> > The only commands that natively open the qf/loc are `:copen`, `:lopen`, `:cwindow`, and `:lwindow`. None of the other qf-related commands do and Vim-qf's raison d'être is precisely to open the qf/loc window after _any_ of those non-opening qf commands is performed.
> 
> IMHO, it seems vim-qf does _much_ more than this. When I'm reading the README
> file, it seems you are referring to this point:
> 
> > * (optional) open the location/quickfix window automatically after :make, :grep, :lvimgrep and friends if there are valid locations/errors
> 
> Thus, it seems perhaps the current way to make vim-qf work with VimTeX is to
> add `let g:qf_auto_open_quickfix = 0`.

I guess I could expose a `g:qf_ignored_events` list and let users blacklist events but I would certainly prefer other plugin authors to restrict themselves to populating the qf/loc list and nothing else.

> > I would consider `:help :caddexpr` as an alternative, as it lets you do all the massaging you need beforehand:
> 
> Thanks for the suggestion, but this does not allow me to use the `errorformat` for parsing the error file? Or am I missing something?

It does. `:caddexpr` is just a "add" variant of `:cexpr`, which parses the given string or list according to `efm`.

A simple example:

    :caddexpr ['~/.bashrc:23:12 first error', '~/.tigrc:10:5 second error']

![caddexpr](https://user-images.githubusercontent.com/344335/114274427-9ede5780-9a1e-11eb-882e-e0334b6c2949.gif)

- Since it is only a simple "quickfix command", it is obviously perfectly compatible with vim-qf or any "open the qf/loc automatically" snippet found on the internet. No stepping over boundaries: your plugin populates the qf/loc list and that's it.
- Since the command expects an expression, it puts you in the driving seat, giving you full control over what is passed to `efm`. Here I pass a list of very simple Unix-like errors which Vim has no problem parsing.

**Comment by lervag** [comment]
> My opinion is that your plugin shouldn't use either `:cwindow` or `:cclose` or make any attempt whatsoever to manipulate the qf/loc window. vimtex runs the current file through some compiler/linter and populates/updates the qf list with the output or a massaged version of the output and that's it. Opening or closing the qf window has nothing to do with LaTeX so vimtex shouldn't do it.

Thanks, I appreciate your opinion, and I think I finally understand what you are saying and even agree with your sentiment. However, as a pragmatic, I will probably not change this within VimTeX just now as I think the current behaviour is appreciated and expected by a lot of users.

> > Thanks for the suggestion, but this does not allow me to use the `errorformat` for parsing the error file? Or am I missing something?
> 
> It does. `:caddexpr` is just a "add" variant of `:cexpr`, which parses the given string or list according to `efm`.
> 
> A simple example:

TIL and all that; thanks again. I agree, it seems `:caddexpr` will allow an even better implementation of the error parsing in VimTeX, and if/when I get the time I will probably use this instead of `:caddfile` and the additional "hacks". However, I notice that `:caddexpr` will also reach the same issue we are discussing above. So:

> I guess I could expose a `g:qf_ignored_events` list and let users blacklist events but I would certainly prefer other plugin authors to restrict themselves to populating the qf/loc list and nothing else.

Although I understand that you might not like this solution very much, I would appreciate it as I think it would allow a simple way to make VimTeX work well with vim-qf. If you don't want it, I'll respect that, and it seems the fallback solution is then to disable the auto-open feature of vim-qf with `g:qf_auto_open_quickfix`.

**Comment by romainl** [comment]
> > My opinion is that your plugin shouldn't use either `:cwindow` or `:cclose` or make any attempt whatsoever to manipulate the qf/loc window. vimtex runs the current file through some compiler/linter and populates/updates the qf list with the output or a massaged version of the output and that's it. Opening or closing the qf window has nothing to do with LaTeX so vimtex shouldn't do it.
> 
> Thanks, I appreciate your opinion, and I think I finally understand what you are saying and even agree with your sentiment. However, as a pragmatic, I will probably not change this within VimTeX just now as I think the current behaviour is appreciated and expected by a lot of users.
> 
> > > Thanks for the suggestion, but this does not allow me to use the `errorformat` for parsing the error file? Or am I missing something?
> > 
> > 
> > It does. `:caddexpr` is just a "add" variant of `:cexpr`, which parses the given string or list according to `efm`.
> > A simple example:
> 
> TIL and all that; thanks again. I agree, it seems `:caddexpr` will allow an even better implementation of the error parsing in VimTeX, and if/when I get the time I will probably use this instead of `:caddfile` and the additional "hacks". However, I notice that `:caddexpr` will also reach the same issue we are discussing above. So:
> 
> > I guess I could expose a `g:qf_ignored_events` list and let users blacklist events but I would certainly prefer other plugin authors to restrict themselves to populating the qf/loc list and nothing else.
> 
> Although I understand that you might not like this solution very much, I would appreciate it as I think it would allow a simple way to make VimTeX work well with vim-qf. If you don't want it, I'll respect that, and it seems the fallback solution is then to disable the auto-open feature of vim-qf with `g:qf_auto_open_quickfix`.

If you take a look at this second gif, which starts where the first one ends, you will notice that a series of `:caddexpr` (but that would also be true for a series of `:caddfile`) works perfectly fine with vim-qf. The behaviour is exactly the one you would expect if you had both vim-qf and a well-behaved LaTeX plugin: said plugin populates the qf list and then vim-qf automatically works its magic. *The pair works perfectly well together*.

![caddexpr2](https://user-images.githubusercontent.com/344335/114281607-89c5f080-9a3f-11eb-93fa-7ea4a7426a95.gif)

If all vimtex does is really `:caddfile` or `:caddexpr` then I have no idea how it is even possible that "vim-qf breaks vimtex" or "vimtex breaks vim-qf". To be frank, I find this whole thread baffling.

**Comment by lervag** [comment]
> If all vimtex does is really `:caddfile` or `:caddexpr` then I have no idea how it is even possible that "vim-qf breaks vimtex" or "vimtex breaks vim-qf".

Well, after more thought I am starting to realize the problem here: I've falsely assumed that `:caddfile` can be used under the assumption that it will never open and jump to the quickfix window. This assumption holds true for the default behaviour, but as you say, the purpose of vim-qf is to open and modify the behaviour of the quickfix window under various conditions; `:caddfile` is just exactly one such example. Thus, my assumption was clearly wrong, and the conclusion is that "vimtex breaks vim-qf".

> To be frank, I find this whole thread baffling.

For me, the thread is enlightening. But I can see how this thread is also _baffling_: It took some time for me to understand that some of my initial assumptions about commands like `:caddexpr` and `:caddfile` were wrong.

Again, thanks for clarifying and explaining things. In my opinion, the thread is now resolved and no further action is required.

**Comment by krishnakumarg1984** [comment]
Err.. Sorry for asking. The technical details of this thread were not easy to understand for me.

Can either of you @romainl or @lervag please provide a quick summary of what the users must do to have both these plugins work smoothly?

**Comment by lervag** [comment]
Nothing; I've updated VimTeX now so I think it should no longer misbehave.

**Comment by krishnakumarg1984** [comment]
Cool. So, there is no need to set `let g:qf_auto_open_quickfix = 0`?

**Comment by lervag** [comment]
Nope, I don't think so. Let me know if you find any more issues.
